### PR TITLE
[LoopVectorize] Improve algorithm for hoisting runtime checks

### PR DESCRIPTION
**Pull request by david-arm**
When attempting to hoist runtime checks out of a loop we currently avoid creating pointer diff checks and prefer to do expanded range checks instead. This gives us the opportunity to hoist runtime checks out of a loop, since these checks are loop invariant. However, in some cases the pointer diff checks would also be loop invariant and so will naturally get hoisted. Therefore, since diff checks are cheaper so we should prefer to use those instead.

### CODE REVIEW

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 1]
```diff
@@ -346,7 +346,10 @@ void RuntimePointerChecking::tryToCreateDiffCheck(
     auto *SinkStartAR = cast<SCEVAddRecExpr>(SinkStartInt);
     const Loop *StartARLoop = SrcStartAR->getLoop();
     if (StartARLoop == SinkStartAR->getLoop() &&
-        StartARLoop == InnerLoop->getParentLoop()) {
+        StartARLoop == InnerLoop->getParentLoop() &&
+        !SE->isKnownPredicate(ICmpInst::ICMP_EQ,
+                              SrcStartAR->getStepRecurrence(*SE),
+                              SinkStartAR->getStepRecurrence(*SE))) {
```
Comment (nikic) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:352]:
```suggestion
        SrcStartAR->getStepRecurrence(*SE) != SinkStartAR->getStepRecurrence(*SE)) {
```
Don't use isKnownPredicate() for ICMP_EQ.

Comment (nikic) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Yeah, SCEVs are uniqued, so a SCEV with the same args will have the same pointer. (Of course, it's possible that expressions are equal in ways that SCEV doesn't understand, but we'd consider that a canonicalization failure.)

**Comment by david-arm** [comment]
The clang formatting errors have nothing to do with my patch - they all seem to complain about existing code.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 2]
```diff
@@ -346,7 +346,9 @@ void RuntimePointerChecking::tryToCreateDiffCheck(
     auto *SinkStartAR = cast<SCEVAddRecExpr>(SinkStartInt);
     const Loop *StartARLoop = SrcStartAR->getLoop();
     if (StartARLoop == SinkStartAR->getLoop() &&
-        StartARLoop == InnerLoop->getParentLoop()) {
+        StartARLoop == InnerLoop->getParentLoop() &&
+        SrcStartAR->getStepRecurrence(*SE) !=
```
Comment (fhahn) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:350]: Would be good to add a comment here explaining the check

**Review by fhahn — COMMENTED** [review]
The test is also failing when I check out the branch for the PR.

It looks like the branch is based on d4cec1ce7313608670d5a97d7c9fcccad68ac659, which doesn't have the diff check improvements I think, which is probably why it is failing.

**Comment by david-arm** [comment]
> The test is also failing when I check out the branch for the PR.
> 
> It looks like the branch is based on [d4cec1c](https://github.com/llvm/llvm-project/commit/d4cec1ce7313608670d5a97d7c9fcccad68ac659), which doesn't have the diff check improvements I think, which is probably why it is failing.

OK thanks I'll try rebasing!

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 2 above)
Comment (david-arm) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:351]: Hi @fhahn, sorry I only just saw this now! I find with github it's really easy to miss review comments compared to Phabricator...

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 3]
```diff
@@ -347,7 +347,12 @@ void RuntimePointerChecking::tryToCreateDiffCheck(
     auto *SinkStartAR = cast<SCEVAddRecExpr>(SinkStartInt);
     const Loop *StartARLoop = SrcStartAR->getLoop();
     if (StartARLoop == SinkStartAR->getLoop() &&
-        StartARLoop == InnerLoop->getParentLoop()) {
+        StartARLoop == InnerLoop->getParentLoop() &&
+        // If the diff check would already be loop invariant (due to the
+        // recurrences being the same), then we should still prefer the diff
```
Comment (paulwalker-arm) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:352]: Perhaps "...,then we prefer to keep the diff check because they are cheaper."

**Review by paulwalker-arm — APPROVED** [review]

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:354]: Yeah the UI does unfortunately sometimes hide some comments :(

**Review by fhahn — APPROVED** [review]
LGTM, thanks for the update!

**Pull request merged**